bpf/lib/policy.h

```diff
@@ -69,13 +69,7 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
 # endif
 
 # if defined(ENABLE_ICMP_RULE)
-			/* Convert from unsigned char to unsigned short
-			 * considering byte order(little-endian).
-			 * In the little-endian case, for example, 2byte data "AB"
-			 * convert to "BA".
-			 * Therefore, the "icmp_type" should be shifted not just casting.
-			 */
-			key.dport = (__u16)(icmphdr.type << 8);
+			key.dport = bpf_htons(icmphdr.type);
```

Comment:
This looks not correct. You are loading icmp hdr from the packet and converting it with htons. It should say bpf_ntohs(), no?

Reply:
I'm also a bit confused. I'm not sure if it's the comment that's incorrect or the code. The key to understanding this will be to confirm exactly where in the dport field the ICMP type is populated by userspace and ensure the datapath matches this.

I'm wondering if it's just the comment that's wrong, which is to say that the encoding of the ICMP type into the dport field has nothing to do with endianness, it's just that the "upper 8 bits" of the 16-bit word is where the ICMP type lives, so the code must do a bitshift.

Reply:
Looks like the ICMP type is parsed in and held in userspace in the host byte-order 16-bit dport field:

https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/policy/api/icmp.go#L208-L211

Which then gets converted to network byte-order (big-endian) when pushed into the datapath:

https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/maps/policymap/policymap.go#L279

If we consider the in-memory layout of the NBO dport field for a policy applying to ICMP type `XX`, we'd expect it to be `00 XX`.

So what we need is that the u8 gets extended to u16 but with (a) big endian: lower 8 bits set to the type value and (b) little endian: upper 8 bits set to the type value.

I agree it's a bit misleading to do this byteswap with `bpf_htons()` given that the value does not start in host byte-order (how could it? It's only 8 bits), but likewise it's a bit misleading to byteswap with `bpf_ntohs()` because the target is not host byte-order. If we break this down, the u8 gets copied into a host byte order u16 variable, so doing `htons()` is more accurate?

Reply:
Thank you for your review!

> so doing htons() is more accurate?

Or should we keep the code and fix the comments?
As you say, using `bpf_htons()` may confuse other developers.

Reply:
I think that keeping `<< 8` is incorrect on big-endian systems, because the value in userspace is stored in big-endian with the lowest byte populated, transferred to the datapath with a no-op byteswap because the system is big-endian, so you must perform a lookup with a big-endian u16 with the lower byte populated. If you read the 8-bit type field into a big-endian u16 then shift it up by 8 bits, you will end up performing a lookup with the _upper_ 8 bits populated. The lookup would then fail.

Reply:
Assuming that we confirm my reasoning above, an alternative could be to just define a new macro in `bpf/lib/endian.h` something like the below, and use that instead:

```
#define bpf_u8_to_be16(x) bpf_htons(x)
```

With this and a well-worded commit message, I think it should be fairly clear for future developers.

Reply:
I see. It seems that we have to use something like `htons()` not shift.

> define a new macro in bpf/lib/endian.h something like the below, and use that instead

It looks nice to me!

Reply:
Throwing another suggestion:
```suggestion
			key.dport = bpf_htons((__u16)icmphdr.type);
```
Casting explicitly instead of relying on the implicit conversion convey the intent better IMHO.

Reply:
Yeah cast looks a bit more explicit :+1: